orderbook/src/database.rs

```diff
@@ -3,14 +3,69 @@ mod fees;
 mod orders;
 mod trades;
 
-use anyhow::Result;
+use anyhow::{Context, Result};
+use chrono::{DateTime, Utc};
+use contracts::gpv2_settlement::Event as ContractEvent;
+use ethcontract::Event as EthContractEvent;
+use futures::{stream::BoxStream, StreamExt};
+use model::{
+    order::{Order, OrderKind, OrderUid},
+    trade::Trade as ModelTrade,
+};
+use primitive_types::{H160, U256};
+use shared::{event_handling::EventIndex, maintenance::Maintaining};
 use sqlx::{Executor, PgPool, Row};
 use std::collections::HashMap;
 
 pub use events::*;
 pub use orders::OrderFilter;
 pub use trades::TradeFilter;
 
+use crate::fee::MinFeeStoring;
+
+#[async_trait::async_trait]
+pub trait Database: Send + Sync {
+    async fn clear(&self) -> Result<()>;
+    async fn count_rows_in_tables(&self) -> Result<HashMap<&'static str, i64>>;
+
+    async fn block_number_of_most_recent_event(&self) -> Result<u64>;
+    async fn insert_events(&self, events: Vec<(EventIndex, Event)>) -> Result<()>;
+    async fn replace_events(
+        &self,
+        delete_from_block_number: u64,
+        events: Vec<(EventIndex, Event)>,
+    ) -> Result<()>;
+    fn contract_to_db_events(
+        &self,
+        contract_events: Vec<EthContractEvent<ContractEvent>>,
+    ) -> Result<Vec<(EventIndex, Event)>>;
+
+    async fn insert_order(&self, order: &Order) -> Result<(), InsertionError>;
+    async fn cancel_order(&self, order_uid: &OrderUid, now: DateTime<Utc>) -> Result<()>;
+    fn orders<'a>(&'a self, filter: &'a OrderFilter) -> BoxStream<'a, Result<Order>>;
+
+    fn trades<'a>(&'a self, filter: &'a TradeFilter) -> BoxStream<'a, Result<ModelTrade>>;
+
+    async fn save_fee_measurement(
```

Comment:
Could `Database: Send + Sync + MinFeeStoring` to avoid redefining these methods? (not chosing to do so might be related to Ben's suggestion of composing the `Database` trait from smaller sub-traits, e.g. an `EventStoring` trait)